### PR TITLE
Better resample

### LINKED ISSUE #80: Resample to seasons - exclude seasons with missing months
### Addressing a Problem?

Currently, resampling data to seasons (e.g., with frequency = 'QS-DEC') can result in partial seasons being counted the same as full seasons. For example, with data for Jan 1980-Dec 1981, Jan 1980 and Dec 1981 will each count as one DJF period even though they are only one month. The DJF mean will then = [Dec 1980 + (Dec 1980-Feb 1981) + Dec 1981]/3. For climatological calculations, Jan 1980-Dec 1981 should probably only give one DJF period (Dec 1980-Feb 1981) but two periods for the other seasons. Having different numbers of periods for each of the seasons can complicate parameter selection for xs.climatological_mean().

### Potential Solution

- My current workaround in this case is to select period = ['1980-03-01', '1981-11-30'] and then set min_periods = window-1 
`ds_clim = xs.climatological_mean(ds=ds_dict['QS-DEC'],  window=2,  interval=2, min_periods = 1, to_level="2yr-climatology")`
- Ideally, any resample to QS-DEC during xs.extract_dataset() would have a "min_periods" type of parameter that refers to a minimum number of months rather than years, or a True/False parameter to only include complete seasons in calculations. 
- xs.climatological_mean() could also be modified to facilitate seasonal calculations since these are fairly common in climatology. 

### Additional context

One challenge is that the metadata generated is largely based on the number if _years_ (e.g., "2yr-climatology") which might be a bit misleading when there is 1 fewer DJF period than the other seasons. Perhaps the metadata can be adapted to take this into account for seasonal calculations.

### LINKED ISSUE #93: Period-length ponderation in means
### Addressing a Problem?

When computing a seasonal mean from monthly data (for example), xarray and xclim will not take the length of each month into account. 

### Potential Solution

Something somewhere to "resample" monthly data to a coarser timescale with month-length weights.

### Additional context

I'm not sure where this should be. 
I noted "climatological_mean" on my sheet, but isn't that function meant for interannual means? Here, the seasonal resampling is a step that comes before the interannual mean.

In #80, we noted that `extract_dataset` was not meant to perform resamples coarser then daily. However, it wouldn't be hard to add a provision in the resampling function to take the period-length into account with periods coarser that a week.

### Contribution

- [ ] I would be willing/able to open a Pull Request to contribute this feature.

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #80 and fixes #93.
- [x] (If applicable) Documentation has been added / updated (for bug fixes / features).
- [x] (If applicable) Tests have been added.
- [x] This PR does not seem to break the templates.
- [x] HISTORY.rst has been updated (with summary of main changes).
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.

### What kind of change does this PR introduce?
- Extends `resample` to explicitly cover the case where the `initial_frequency` is coarser than "W", meaning that it has a non-uniform length.
    + Weights are the proportion of days of the total period contributed by each timesteps. Ex: MS to YS => weight = daysinmonth / 365 .
    + "weighted resample" is activated for methods : "wind_direction", "mean", "std", "var" and "median". Those are the methods from `DataArray.resample` for which a weighted version makes sense.
    + Implemented a generic version that uses `resample(...).map(....weighted().op()) `, with a special case for "median" since that's not a `weighted` op. Mapping a function is less performant, especially since flox can't be used.
    + Implemented a shortcut for `method='mean'` that simplifies to a simple `resample().sum()`, meaning it should be much more performant and able to use `flox`. Without dask and on the air dataset, I had a 3x speedup.

- Extends `resample` to handle missing values with new `missing` arg:
    + `mask` : Mask incomplete periods (as in missing timesteps)
    + `drop`: Drop incomplete periods (same)
    + dict with `method=<xclim missing method>`  : Mask periods failing specified xclim missing check.

### Does this PR introduce a breaking change?
No.

### CODE REVIEW

File: xscen/extract.py [hunk 1]
```diff
@@ -368,45 +368,73 @@ def resample(
...
+        The target frequency/freq str, must be one of the frequency supported by xarray.
     ds : xr.Dataset, optional
         The "wind_direction" resampling method needs extra variables, which can be given here.
     method : {'mean', 'min', 'max', 'sum', 'wind_direction'}, optional
         The resampling method. If None (default), it is guessed from the variable name and frequency,
         using the mapping in CVs/resampling_methods.json. If the variable is not found there,
         "mean" is used by default.
+    missing: {'mask', 'drop'} or dict, optional
+        If 'mask' or 'drop', target periods with fewer steps than expected are masked or dropped.
+        For example, for daily data beginning in January with a `target_frequency` of "QS-DEC". the first season is missing one month.
```
Comment (RondeauG) [inline xscen/extract.py:393]:
```suggestion
        If 'mask' or 'drop', target periods that would have been computed from fewer timesteps than expected are masked or dropped, using a threshold of 5% of missing data.
        For example, the first season of a `target_frequency` of "QS-DEC" will be masked or dropped if data only starts in January.
```

**Review by RondeauG — APPROVED** [review]
LGTM!

File: xscen/extract.py [hunk 2]
```diff
@@ -368,45 +368,73 @@ def resample(
...
     )
-    if initial_frequency_td == pd.Timedelta("1D"):
-        logger.warning(
-            "You appear to be resampling daily data using extract_dataset. "
-            "It is advised to use compute_indicators instead, as it is far more robust."
+
+    weights = None
+    if (
+        initial_frequency != "undetected"
+        and initial_frequency_td > pd.Timedelta(7, "days")
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/extract.py]: Indeed that could be an issue! Last commit implements a better `xrfreq_to_timedelta`.

**Review by mccrayc — COMMENTED** [review]
Looks good, thanks!

**Pull request merged**